cpp/velox/substrait/SubstraitToVeloxPlan.h

```diff
@@ -220,23 +220,33 @@ class SubstraitToVeloxPlanConverter {
           rightBound_ = true;
         return !leftBound_;
       }
-      if (rightBound_ || inRange_ || multiRange_) {
+      if (rightBound_ || inRange_ || multiRange_ || isNull_) {
         return false;
       }
       rightBound_ = true;
       return true;
     }
 
-    /// Set the multi-range and returns whether it can coexist with
+    /// Set the existence of multi-range and returns whether it can coexist with
     /// existing conditions for this field.
     bool setMultiRange() {
-      if (inRange_ || multiRange_ || leftBound_ || rightBound_) {
+      if (inRange_ || multiRange_ || leftBound_ || rightBound_ || isNull_) {
         return false;
       }
       multiRange_ = true;
       return true;
     }
 
+    /// Set the existence of IsNull and returns whether it can coexist with
+    /// existing conditions for this field.
+    bool setIsNull() {
+      if (inRange_ || multiRange_ || leftBound_ || rightBound_) {
```

Comment:
Yes, only one of them can be pushed down because IsNull cannot coexist with other range filter for a single field. But I doubt whether Spark generates such conditions. For query like `select l_returnflag from lineitem where l_returnflag > 1 and l_returnflag is null`, Spark only produces IsNotNull and IsNull conditions.
> (1) NativeScan parquet 
Output [1]: [l_returnflag#8]
Batched: true
Location: InMemoryFileIndex [hdfs://sr245:9001/tpch_date_10/lineitem]
PushedFilters: [IsNotNull(l_returnflag), IsNull(l_returnflag)]
ReadSchema: struct<l_returnflag:string>

Reply:
@rui-mo The datatype for `l_returnflag` is string. If replacing `l_returnflag` with `l_orderkey` using Vanilla spark:
`spark.sql("select l_returnflag from lineitem where l_orderkey > 1 and l_orderkey is null")`

`df.explain()` prints
```
== Physical Plan ==
*(1) Project [l_returnflag#8]
+- *(1) Filter ((isnotnull(l_orderkey#0L) AND (l_orderkey#0L > 1)) AND isnull(l_orderkey#0L))
   +- *(1) ColumnarToRow
      +- FileScan parquet [l_orderkey#0L,l_returnflag#8] Batched: true, DataFilters: [isnotnull(l_orderkey#0L), (l_orderkey#0L > 1), isnull(l_orderkey#0L)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[hdfs://user-S520-B3.sh.intel.com:8020/tpch_sf2000_parquet_snappy/linei..., PartitionFilters: [], PushedFilters: [IsNotNull(l_orderkey), GreaterThan(l_orderkey,1), IsNull(l_orderkey)], ReadSchema: struct<l_orderkey:bigint,l_returnflag:string>
```

And `df.collect()` prints
```
res11: Array[org.apache.spark.sql.Row] = Array()
```

It looks like both "IsNull" and "GreaterThan" are pushed down.

Reply:
OK. Good to know!